api/dataset.go

```diff
@@ -382,6 +382,132 @@ func (api *DatasetAPI) addVersion(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+
+	dataset, err := models.CreateDataset(r.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	defer r.Body.Close()
+
+	if err := api.dataStore.Backend.UpdateDataset(datasetID, dataset); err != nil {
+		handleErrorType(err, w)
+		return
+	}
+
+	setJSONContentType(w)
+	w.WriteHeader(http.StatusOK)
+	log.Debug("update dataset", log.Data{"dataset_id": datasetID})
+}
+
+func (api *DatasetAPI) putVersion(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	datasetID := vars["id"]
+	editionID := vars["edition"]
+	versionID := vars["version"]
+
+	if r.Header.Get(internalToken) != api.internalToken {
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+
+	version, err := models.CreateVersion(r.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	defer r.Body.Close()
+
+	currentVersion, err := api.dataStore.Backend.GetVersion(datasetID, editionID, versionID, "")
+	if err != nil {
+		handleErrorType(err, w)
+		return
+	}
+
+	// Check current state of version document;
+	// if published do not try to update document
+	if currentVersion.State == publishedState {
+		http.Error(w, fmt.Sprintf("Unable to update document, already published"), http.StatusForbidden)
+		return
+	}
+
+	// Combine update version document to existing version document
+	newVersion := createNewVersionDoc(currentVersion, version)
+
+	if err = models.ValidateVersion(newVersion); err != nil {
+		log.ErrorR(r, err, nil)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if err := api.dataStore.Backend.UpdateVersion(newVersion.ID, version); err != nil {
+		handleErrorType(err, w)
+		return
+	}
+
+	if version.State == publishedState {
+		if err := api.dataStore.Backend.UpdateEdition(newVersion.Links.Edition.ID, version.State); err != nil {
+			log.ErrorC("failed to update the state of edition document to published", err, nil)
+			handleErrorType(err, w)
+			return
+		}
+
+		// Pass in newVersion variable to include relevant data needed for update on dataset API (e.g. links)
+		if err := api.updateDataset(datasetID, newVersion); err != nil {
+			log.ErrorC("failed to update dataset document once version state changes to publish", err, nil)
+			handleErrorType(err, w)
+			return
+		}
+	}
+
+	if version.State == associatedState {
+		if err := api.dataStore.Backend.UpdateDatasetWithAssociation(datasetID, associatedState, version); err != nil {
+			log.ErrorC("failed to update dataset document after a version of a dataset has been associated with a collection", err, nil)
+			handleErrorType(err, w)
+			return
+		}
+	}
+
+	setJSONContentType(w)
+	w.WriteHeader(http.StatusOK)
+	log.Debug("update dataset", log.Data{"dataset_id": datasetID})
+}
+
+func createNewVersionDoc(currentVersion *models.Version, version *models.Version) *models.Version {
+	if version.CollectionID == "" && currentVersion.CollectionID != "" {
```

Comment:
Consider changing the logic here to check if the value has changed, instead of just gone from empty to not empty. e.g. `if version.CollectionID != currentVersion.CollectionID`

Reply:
Its a bit more subtle then just empty to not empty. It literally means do we have a `collection_id` in the version doc we were sent if not then do we have one in the current version doc to add to the version doc; if we do then only in this scenario do we set the `version.CollectionID` to the `currentVersion.CollectionID`.

Reply:
The point I was making is that an existing value != "" would not be overwritten. Only an empty existing value would be set. You may not want to update the value if its empty. A more complete example to demonstrate: 
`if version.CollectionID != "" && if version.CollectionID != currentVersion.CollectionID`